pkg/services/dashboards/database/database_dashboard_public.go

```diff
@@ -0,0 +1,159 @@
+package database
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/util"
+	"github.com/grafana/grafana/pkg/util/errutil"
+)
+
+// retrieves public dashboard configuration
+func (d *DashboardStore) GetPublicDashboard(uid string) (*models.PublicDashboard, *models.Dashboard, error) {
+	if uid == "" {
+		return nil, nil, models.ErrPublicDashboardIdentifierNotSet
+	}
+
+	// get public dashboard
+	pdRes := &models.PublicDashboard{Uid: uid}
+	err := d.sqlStore.WithTransactionalDbSession(context.Background(), func(sess *sqlstore.DBSession) error {
+		has, err := sess.Get(pdRes)
+		if err != nil {
+			return err
+		}
+		if !has {
+			return models.ErrPublicDashboardNotFound
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// find dashboard
+	dashRes := &models.Dashboard{OrgId: pdRes.OrgId, Uid: pdRes.DashboardUid}
+	err = d.sqlStore.WithTransactionalDbSession(context.Background(), func(sess *sqlstore.DBSession) error {
+		has, err := sess.Get(dashRes)
+		if err != nil {
+			return err
+		}
+		if !has {
+			return models.ErrPublicDashboardNotFound
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return pdRes, dashRes, err
+}
+
+// generates a new unique uid to retrieve a public dashboard
+func generateNewPublicDashboardUid(sess *sqlstore.DBSession) (string, error) {
+	for i := 0; i < 3; i++ {
+		uid := util.GenerateShortUID()
+
+		exists, err := sess.Get(&models.PublicDashboard{Uid: uid})
+		if err != nil {
+			return "", err
+		}
+
+		if !exists {
+			return uid, nil
+		}
+	}
+
+	return "", models.ErrPublicDashboardFailedGenerateUniqueUid
+}
+
+// retrieves public dashboard configuration
+func (d *DashboardStore) GetPublicDashboardConfig(orgId int64, dashboardUid string) (*models.PublicDashboardConfig, error) {
+	if dashboardUid == "" {
+		return nil, models.ErrDashboardIdentifierNotSet
+	}
+
+	// get global dashboard config
+	dashRes := &models.Dashboard{OrgId: orgId, Uid: dashboardUid}
+	err := d.sqlStore.WithTransactionalDbSession(context.Background(), func(sess *sqlstore.DBSession) error {
+		has, err := sess.Get(dashRes)
+		if err != nil {
+			return err
+		}
+		if !has {
+			return models.ErrDashboardNotFound
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	// get public dashboards
+	pdRes := &models.PublicDashboard{OrgId: orgId, DashboardUid: dashboardUid}
+	err = d.sqlStore.WithTransactionalDbSession(context.Background(), func(sess *sqlstore.DBSession) error {
+		_, err := sess.Get(pdRes)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	pdc := &models.PublicDashboardConfig{
+		IsPublic:        dashRes.IsPublic,
+		PublicDashboard: *pdRes,
+	}
```

Comment:
the dashRes.IsPublic property seems unnecessary here, the dashboard is public if there exists a dashboard_public_config, I think the is_public column on the dashboard can be removed, seems to be superfluous? I might be missing a the use case
why this flag exists.

Reply:
It's just the messy nesting and having to update the main dashboard feels so clunky, and the nestings leads to if statements like this :) 
`if len(cmd.PublicDashboardConfig.PublicDashboard.DashboardUid) == 0 {` 

In what way can one dashboard have many public dashboards? Do you mean they can have multiple access keys with different time ranges and variable settings? 

I am starting to understand, this one-to-many design was not at all clear to me. It's still not super clear. Especially the whole access key + settings (time range + variables), and how that is thought to be scoped/shared. 

We are working on new dashboard file-based storage systems (and git integration). Similar to provisioning but where the dashboards will not exist in the DB but only in the external store. So having to update a dashboard in git or a provisioned file to make it public will be messy. Same with dashboards that come from integrations that are read-only. Having it self-contained in dashboard_public_config would simplify things and make it possible for multiple access keys to share the same settings, might require another table however to support the one-to-many bit. 

Anyway I am really sorry for all my comments, and the slow feedback cycle.